main/src/library/MutDisjointSet.flix

```diff
@@ -0,0 +1,160 @@
+        if (not memberOf(x, s)) {
+            MutMap.put!(x, singleton(rc, x), forest)
+        } else {
+            ()
+        }
+    }
+
+    ///
+    /// Returns `true` iff `x` is a member of `s`.
+    ///
+    pub def memberOf(x: t, s: MutDisjointSet[t, r]): Bool \ r with Order[t] = {
+        let MutDisjointSet({ forest | _ }) = s;
+        MutMap.memberOf(x, forest)
+    }
+
+    ///
+    /// Returns a new `Node` from `x` in `rc`.
+    ///
+    def singleton(rc: Region[r], x: t): Node[t, r] \ r =
+        Node.Node({ rc = rc, value = x, parent = Ref.new(rc, None), rank = Ref.new(rc, 0) })
+
+    ///
+    /// Updates `s` with the elements of the iterable `m`.
+    ///
+    pub def makeSets(m: m[t], s: MutDisjointSet[t, r]): Unit \ r with Iterable[m], Order[t] = region rc {
+        Iterable.iterator(rc, m) |> Iterator.forEach(x -> makeSet(x, s))
+    }
+
+    ///
+    /// Returns the representative / root of the set that contains `x`.
+    ///
+    pub def find(x: t, s: MutDisjointSet[t, r]): Option[t] \ r with Order[t] =
+        findSet(x, s) |> Option.map(match Node.Node({ value | _ }) -> value)
+
+    ///
+    /// Returns the representative / root `Node` of the set that contains `x`.
+    ///
+    def findSet(x: t, s: MutDisjointSet[t, r]): Option[Node[t, r]] \ r with Order[t] = {
+        def findRoot(y) = {
+            let Node.Node({ parent | _ }) = y;
+            match Ref.get(parent) {
+                case Some(p) => {
+                    let root = findRoot(p);
+                    Ref.put(Some(root), parent);
+                    root
+                }
+                case None => y
+            }
+        };
+        let MutDisjointSet({ forest | _ }) = s;
+        MutMap.get(x, forest)
+        |> Option.map(findRoot)
+    }
+
+    ///
+    /// Merges the sets that contain `x` and `y` if both `x` and `y` are in `s`.
+    /// If `x` and `y` have the same rank, then `y` becomes the new parent of `x`.
+    ///
+    pub def union(x: t, y: t, s: MutDisjointSet[t, r]): Unit \ r with Order[t] =
+        def link(x1, y1) = {
+            if (not equal(x1, y1)) {
+                let Node.Node({ parent = xParent, rank = xr | _ }) = x1;
+                let Node.Node({ parent = yParent, rank = yr | _ }) = y1;
+                let xRank = Ref.get(xr);
+                let yRank = Ref.get(yr);
+                if (xRank > yRank) {
+                    Ref.put(Some(x1), yParent)
+                } else {
+                    Ref.put(Some(y1), xParent)
+                };
+                if (xRank == yRank) {
+                    Ref.put(yRank + 1, yr)
+                } else {
+                    ()
+                }
+            }
+            else {
+                ()
+            }
+        };
+        match (findSet(x, s), findSet(y, s)) {
+            case (Some(xn), Some(yn)) => link(xn, yn)
+            case _ => ()
+        }
+
+    ///
+    /// Returns true if `x` and `y` have the same representative / root.
```

Comment:
perhaps we should say "are in the same disjoint set"?

Reply:
I think that's a good idea, but I think we should keep both, since readers familiar with the literature will know what this means.

Reply:
OK. lets have both.